controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +279,134 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol)
+		if err != nil {
+			return err
+		}
+		if isVolumeInRebuilding {
+			return nil
+		}
+	}
 
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuilding longhorn.VolumeOfflineRebuilding) (bool, error) {
+	if offlineRebuilding == longhorn.VolumeOfflineRebuildingEnabled {
+		return true, nil
+	}
+
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return offlineReplicaRebuildingValue && offlineRebuilding != longhorn.VolumeOfflineRebuildingDisabled, nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachmentForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	areAllReplicasHealthy, err := vbc.isVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas)
+	if err != nil {
+		return va, err
+	}
+	if !areAllReplicasHealthy {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.AnyValue, longhorn.AttacherTypeVolumeRebuildingController)
+	}
+	return va, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas int, replicas map[string]*longhorn.Replica) (bool, error) {
+	if len(replicas) < numberOfReplicas {
+		return false, nil
+	}
+
+	healthyCount := 0
+	for _, replica := range replicas {
+		node, err := vbc.ds.GetNodeRO(replica.Spec.NodeID)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				// requeue the volume if getting node fails
+				return false, err
+			}
+			// If the node is not found, it means the replica is not healthy
+			continue
+		}
+		if node.Status.Conditions == nil {
+			continue
+		}
+		nodeReadyCondition := types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady)
+		if nodeReadyCondition.Status != longhorn.ConditionStatusTrue {
+			continue
+		}
+
+		if replica.Spec.FailedAt != "" {
+			continue
+		}
+		// the volume might be newly created and not attached yet.
+		if replica.Spec.HealthyAt != "" || (replica.Spec.LastFailedAt == "" && replica.Spec.LastHealthyAt == "") {
+			healthyCount++
+		}
+	}
+	return healthyCount >= numberOfReplicas, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasRebuilding(vol *longhorn.Volume) (bool, error) {
+	engines, err := vbc.ds.ListVolumeEngines(vol.Name)
+	if err != nil {
+		return false, err
+	}
+	engine, err := vbc.ds.PickVolumeCurrentEngine(vol, engines)
+	if err != nil {
+		return false, err
+	}
+	if engine == nil || engine.Status.ReplicaModeMap == nil {
```

Comment:
Why return true while engine == nil?

Reply:
Moved out the logic. And yes, it should not be `nil` unless an unexpected error.
Thanks.